# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.h
# Commit: a86ecdaa1716
# Full Hash: a86ecdaa1716198e2aa2b60573fc5b276fc22b98
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-06-30 04:06:43
# Regressor Bug: 1831451
# File Overlap Count: 1
# Description:
#   Bug 1831451 - Fire a dequeue event when data is being decoded r=padenot
#   
#   This patch implements the dequeue-event dispatch [1] that decode method
#   and reset method need.
#   
# ==============================================================================

diff -r 226dda36eb9e -r a86ecdaa1716 dom/media/webcodecs/VideoDecoder.h
--- a/dom/media/webcodecs/VideoDecoder.h	Thu Jun 29 22:33:12 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.h	Thu Jun 29 22:33:12 2023 +0000
@@ -79,6 +79,8 @@
   NS_DECL_ISUPPORTS_INHERITED
   NS_DECL_CYCLE_COLLECTION_CLASS_INHERITED(VideoDecoder, DOMEventTargetHelper)
 
+  IMPL_EVENT_HANDLER(dequeue)
+
  public:
   VideoDecoder(nsIGlobalObject* aParent,
                RefPtr<WebCodecsErrorCallback>&& aErrorCallback,
@@ -99,10 +101,6 @@
 
   uint32_t DecodeQueueSize() const;
 
-  already_AddRefed<EventHandlerNonNull> GetOndequeue() const;
-
-  void SetOndequeue(EventHandlerNonNull* arg);
-
   MOZ_CAN_RUN_SCRIPT void Configure(const VideoDecoderConfig& aConfig,
                                     ErrorResult& aRv);
 
@@ -135,6 +133,8 @@
 
   void ScheduleClose(const nsresult& aResult);
 
+  void ScheduleDequeueEvent();
+
   MOZ_CAN_RUN_SCRIPT void ProcessControlMessageQueue();
   void CancelPendingControlMessages();
 
@@ -168,6 +168,8 @@
   RefPtr<DecoderAgent> mAgent;
   UniquePtr<VideoDecoderConfig> mActiveConfig;
   uint32_t mDecodeQueueSize;
+  bool mDequeueEventScheduled;
+
   DecodeMessage::Id mDecodeMessageCounter;
 
   // Used to add a nsIAsyncShutdownBlocker on main thread to block